src/cowboy_telemetry_h.erl

```diff
@@ -37,40 +37,56 @@ metrics_callback(#{early_error_time := Time} = Metrics) when is_number(Time) ->
         [cowboy, request, early_error],
         #{system_time => erlang:system_time(), resp_body_length => RespBodyLength},
         Metadata);
-metrics_callback(#{reason := {internal_error, {'EXIT', _, {_, Stacktrace}}, _}} = Metrics) ->
+metrics_callback(#{reason := {internal_error, {'EXIT', _, {Reason, Stacktrace}}, _}} = Metrics) ->
     telemetry:execute(
         [cowboy, request, exception],
         measurements(Metrics),
-        metadata(Metrics, #{kind => exit, stacktrace => Stacktrace}));
+        (metadata(Metrics))#{kind => exit, reason => Reason, stacktrace => Stacktrace});
 metrics_callback(#{reason := {ErrorType, _, _} = Reason} = Metrics)
     when ErrorType == socket_error;
          ErrorType == connection_error ->
     telemetry:execute(
         [cowboy, request, stop],
         measurements(Metrics),
-        metadata(Metrics, #{error => Reason}));
+        (metadata(Metrics))#{error => Reason});
 metrics_callback(Metrics) ->
     telemetry:execute(
         [cowboy, request, stop],
         measurements(Metrics),
-        metadata(Metrics, #{})).
+        metadata(Metrics)).
 
 measurements(Metrics) ->
-    Measurements = maps:with([req_body_length, resp_body_length], Metrics),
-    Durations =
-        #{duration => duration(req_start, req_end, Metrics),
-          req_body_duration => duration(req_body_start, req_body_end, Metrics),
-          resp_duration => duration(resp_start, resp_end, Metrics)},
-    maps:merge(Measurements, Durations).
+    #{req_body_length := ReqBodyLength, resp_body_length := RespBodyLength} = Metrics,
 
-metadata(Metrics, Extra) ->
-    Metadata = maps:with([pid, streamid, req, resp_headers, resp_status, reason, procs, informational, ref], Metrics),
-    maps:merge(Metadata, Extra).
+    #{
+        duration => duration(req_start, req_end, Metrics),
+        req_body_duration => duration(req_body_start, req_body_end, Metrics),
+        resp_duration => duration(resp_start, resp_end, Metrics),
+        req_body_length => ReqBodyLength,
+        resp_body_length => RespBodyLength
+    }.
 
-duration(StartKey, EndKey, Metrics) ->
-    duration(maps:get(StartKey, Metrics, undefined), maps:get(EndKey, Metrics, undefined)).
+metadata(Metrics) ->
+    #{
+        pid := Pid,
+        streamid := Streamid,
+        req := Req,
+        resp_headers := RespHeaders,
+        resp_status := RespStatus,
+        ref := Ref
+    } = Metrics,
+
+    #{
+        pid => Pid,
+        streamid => Streamid,
+        req => Req,
+        resp_headers => RespHeaders,
+        resp_status => RespStatus,
+        ref => Ref
+    }.
```

Comment:
@binaryseed I have reduced the number of keys for now - since some of the metadata themselves were measurements - and to reduce the number of operations. Another alternative is to pass all of the `Metrics` as metadata and call it a day. Do you have a preference?

Reply:
Nice. So I started with that actually, passing everything as metadata. I moved forward with the option of splitting them to be more true to the expectations of `telemetry` events, since some are clearly measurements, and others are pretty clearly metadata.

However, I get that this instrumentation in particular is quite performance sensitive since it'll run so often, so I'm open.

Do you think performance with your current changes will be sufficient?

Reply:
(I do like the changes in this PR)

Reply:
To be fair, I think the code before this PR would also be completely fine but if we can improve things a bit further, then why not? :)

Reply:
Agreed!